programs/util.h

```diff
@@ -191,7 +191,7 @@ UTIL_STATIC void* UTIL_realloc(void* ptr, size_t size)
 {
     void* const newptr = realloc(ptr, size);
     if (newptr) return newptr;
-    free(ptr);
+    if (size!=0) free(ptr);
```

Comment:
I believe it's preferable to do this check at the beginning the function, and then exit early with `return NULL`.
This avoids any risk associated with `realloc(ptr, 0)`, which may or may not act or allocate something, depending on `libc` implementation, and we don't want to be dependent on a specific behavior.